docs/proposals/etcd-druid.md

```diff
@@ -0,0 +1,24 @@
+# Integrating etcd-druid with Gardener.
+
+Etcd is currently deployed by garden-controller-manager as a Statefulset. The sidecar container spec contains details pertaining to cloud-provider object-store which is injected into the statefulset via a mutable webhook running as part of the gardener extension [story](https://github.com/gardener/gardener/blob/master/docs/extensions/controlplane-webhooks.md#what-needs-to-be-implemented-to-support-a-new-cloud-provider).  This approach restricts the operations on etcd such as scale-up and upgrade. Etcd-druid will eliminate the need to hijack statefulset creation to add cloudprovider details. It has been designed to provide an intricate control over the procedure of deploying and maintaining etcd. The roadmap for etcd-druid can be found [here](https://github.com/gardener/etcd-druid/issues/2). etcd-
+
+This document explains how Gardener deploys etcd and what resources it creates for etcd-druid to deploy an etcd cluster.
+
+## Resources required by etcd-druid (created by Gardener)
+
+* Secret containing credentials to access backup bucket in Cloud provider object store.
+* TLS server and client secrets for etcd and backup-sidecar
+* Etcd CRD resource that contains parameters pertaining to etcd, backup-sidecar and cloud-provider object store.
+
+When an etcd resource is created in the cluster, the druid acts on it by creating an etcd statefulset, a service and a configmap containing etcd bootstrap script. The secrets containing the infrastructure credentials and the TLS certificates are mounted as volumes. 
+
+Garden-controller-manager, being cloud agnostic, deploys the etcd resource. This will not contain any cloud-specific information other than the cloud-provider. The extension controller that contains the cloud specific implementation to create the backup bucket will create it if needed and create a secret containing the credentials to access the bucket. The secret will have to be made available in the namespace the etcd statefulset will be deployed. If etcd and backup-sidecar communicates over TLS then the CA certificates, server and client certificates, and keys will also have to be made available in the namespace as well. The etcd resource will have reference to these aforementioned secrets. etcd-druid will deploy the statefulset only if the secrets are available.
```

Comment:
> The secret will have to be made available in the namespace the etcd statefulset will be deployed.

Yes, that's what the `BackupEntry` controller does today. Additionally, I suggest to expose the secret name in the `status` as written above.

> If etcd and backup-sidecar communicates over TLS then the CA certificates, server and client certificates, and keys will also have to be made available in the namespace as well.

Sure, that can and is already done by Gardener. We can keep it as is.

> The etcd resource will have reference to these aforementioned secrets.

This part I don't really understand. Do the extension controllers inject the cloud provider specific information wrt backup via webhook? If not, how gets the information in otherwise?

> etcd-druid will deploy the statefulset only if the secrets are available.

See above question; it would be good if the etcd+backup-restore sidecar would also be deployed if no backup information is given - then, only used for corruptness checks etc.

Reply:
Gardener should provide which cloudprovider would be used to store the backups. Druid will figure out the object store with the cloudprovider as we assure there will be a one-one mapping. Regarding the backup secret, should it be fine that the `etcd-backup` secretRef is added to the etcd resource. Gardener can do that in a cloud-provider agnostic way as the information is in the secret. Right?

Reply:
> Regarding the backup secret, won't it be fine that the etcd-backup secretRef is added to the etcd resource by Gardener.  Gardener can do that in a cloud-provider agnostic way as the information is in the secret. Right?

Yes, this would be okay if the above is implemented (the etcd backup secret name is exposed in the `BackupEntry` status). Then, Gardener can read it and write it into the ETCD resource as you mentioned. Please update the document accordingly.
Also, please prefix the file with `06-` --> `06-etcd-druid.md`.